packages/sync-actions/test/product-sync-variants.spec.js

```diff
@@ -291,42 +291,186 @@ describe('Actions', () => {
-        // changed
-        { id: 2, sku: 'v2', key: 'v2', attributes: [{ name: 'foo', value: 'another value' }] },
-        // changed
-        { id: 3, sku: 'v3', key: 'v3', attributes: [{ name: 'foo', value: 'i dont care' }] },
-        // new
-        { sku: 'v4', key: 'v4', attributes: [{ name: 'foo', value: 'yet another' }] },
-      ],
-    }
+    describe('with variants in `now`', () => {
+      const before = {
+        id: '123',
+        version: 1,
+        variants: [
+          { id: 2, sku: 'v2', key: 'v2', attributes: [{ name: 'foo', value: 'qux' }] },
+          { id: 3, sku: 'v3', key: 'v3', attributes: [{ name: 'foo', value: 'baz' }] },
+        ],
+      }
 
-    const actions = productsSync.buildActions(now, before)
+      const now = {
+        id: '123',
+        variants: [
+          // changed
+          { id: 2, sku: 'v2', key: 'v2', attributes: [{ name: 'foo', value: 'another value' }] },
+          // changed
+          { id: 3, sku: 'v3', key: 'v3', attributes: [{ name: 'foo', value: 'i dont care' }] },
+          // new
+          { sku: 'v4', key: 'v4', attributes: [{ name: 'foo', value: 'yet another' }] },
+        ],
+      }
+
+      it('should generate `addVariant` and `setAttribute` actions',
+      () => {
+        const actions = productsSync.buildActions(now, before)
+
+        expect(actions).toEqual([
+          { action: 'addVariant', sku: 'v4', key: 'v4', attributes: [{ name: 'foo', value: 'yet another' }] },
+          { action: 'setAttribute', variantId: 2, name: 'foo', value: 'another value' },
+          { action: 'setAttribute', variantId: 3, name: 'foo', value: 'i dont care' },
+        ])
+      })
+    })
 
-    expect(actions).toEqual([
-      { action: 'addVariant', sku: 'v4', key: 'v4', attributes: [{ name: 'foo', value: 'yet another' }] },
-      { action: 'setAttribute', variantId: 2, name: 'foo', value: 'another value' },
-      { action: 'setAttribute', variantId: 3, name: 'foo', value: 'i dont care' },
-    ])
+    describe('with changed non master variant in `now`', () => {
+      describe('with non existing variant', () => {
+        const before = {
+          id: '123',
+          version: 1,
+          masterVariant: {
+            id: 1,
+            sku: 'v1',
+            attributes: [{ name: 'foo', value: 'bar' }],
+          },
+          variants: [],
+        }
+
+        const now = {
+          id: '123',
+          version: 1,
+          masterVariant: {
+            id: 2,
+            sku: 'v1',
+            attributes: [{ name: 'foo', value: 'bar' }],
+          },
+          variants: [],
+        }
+
+        it('should generate `changeMasterVariant` and `addVariant` action',
+        () => {
+          const actions = productsSync.buildActions(now, before)
+
+          expect(actions).toEqual([
+            { action: 'addVariant', attributes: [{ name: 'foo', value: 'bar' }], id: 2, sku: 'v1' },
+            { action: 'changeMasterVariant', variantId: 2 },
```

Comment:
It should also delete the variant with id `1`, shouldn't it?

Reply:
This is one of the semantics I wasn't 100% sure about. It felt safer not to remove than to just throw it away. I'll see what other reviewers say.

Reply:
if the goal is that `now` = `before` + all update actions, then we have to remove the variant. Otherwise we have one variant too much in `now`.

Reply:
Yeap. You're right. I am thinking how we can bake it in.

Reply:
Yeap, I agree with us removing the update actions.